### PR TITLE
dusty env command

### LINKED ISSUE #474: RFC: `dusty env` to allow user to specify environment overrides
It can often be useful to change the behavior of an application by manipulating environment variables rather than code. A common example of this would be changing the log level to a more verbose setting in order to debug an issue. Currently, the only ways to do this with Dusty are to override the setting in the code (which doesn't work for services) or override the Dusty specs and override the environment in the `compose` key (which is clumsy and doesn't scale).

I'm proposing the addition of an `env` CLI subcommand which would allow the user to specify environment variable overrides to containers (apps or services) run by Dusty. These overrides would be specific to the user's machine and would not pollute the Dusty specs. Something like:

```
dusty env set <app or service name> <variable name> <variable value>
# e.g.
dusty env set website LOG_LEVEL debug
```

During `dusty up` and similar commands, Dusty would merge the current overrides into any existing environment settings in the specs, with preference being given to the overrides.

We would also need commands to unset the overrides and display to the user which environment overrides are currently registered, e.g.

```
dusty env unset <app or service name> <variable name>
dusty env list <app or service name>
```

### PR DESCRIPTION
**Pull request by jsingle**
@thieman still gotta write tests but it works.  The main code is ready for you to start looking at

Resolves #474

### CODE REVIEW

File: dusty/commands/env.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import functools
```
Comment (jsingle) [inline dusty/commands/env.py]: oops UNUSED

File: dusty/commands/env.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+from prettytable import PrettyTable
+
+from ..log import log_to_client
+from ..compiler.spec_assembler import get_assembled_specs, get_specs
+from ..config import get_env_config, save_config_value
+from .. import constants
+from ..payload import daemon_command
+
+def _save_env(global_env_dict):
+    save_config_value(constants.CONFIG_ENV_KEY, global_env_dict)
```
Comment (thieman) [inline dusty/commands/env.py]: Never use `global` in a variable name unless it's actually a global variable, which this isn't

File: dusty/commands/env.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    for app_or_service, env_dict in global_env_dict.iteritems():
+        if app_or_service in active_containers:
+            for var_name, value in env_dict.iteritems():
+                table.add_row([app_or_service, var_name, value])
+    log_to_client(table.get_string())
+
+@daemon_command
+def set_from_file(app_or_service, local_file_path):
+    _assert_app_or_service_exists(app_or_service)
+    if not os.path.exists(local_file_path):
```
Comment (thieman) [inline dusty/commands/env.py]: This should also check whether the file is readable, I think you can use `os.stat` for that

File: dusty/commands/env.py (hunk 2 above)
Comment (jsingle) [inline dusty/commands/env.py]: haha good point

File: dusty/commands/env.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    if not global_env_dict.get(app_or_service):
+        global_env_dict[app_or_service] = {}
+    with open(local_file_path, 'r') as f:
+        for line in f.readlines():
+            equal_split = line.split('=', 1)
+            if len(equal_split) != 2:
+                raise RuntimeError('Line {} is not of the format VARIABLE=value'.format(line))
+            var_name = equal_split[0]
+            value = equal_split[1].rstrip()
+            global_env_dict[app_or_service][var_name] = value
```
Comment (thieman) [inline dusty/commands/env.py]: Find a parser for this. Env files support comments and stuff.

File: dusty/commands/env.py (hunk 3 above)
Comment (jsingle) [inline dusty/commands/env.py]: K

File: tests/integration/cli/env_test.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    def setUp(self):
+        super(TestEnvCLI, self).setUp()
+        single_specs_fixture()
+        local_dir = Repo('github.com/app/a').managed_path
+        if not os.path.exists(local_dir):
+            os.makedirs(local_dir)
+        self.run_command('bundles activate bundle-a')
+
+    def tearDown(self):
+        self.run_command('stop')
```
Comment (thieman) [inline tests/integration/cli/env_test.py]: Don't we usually put this in a `try`? Can't remember exactly why

File: tests/integration/cli/env_test.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        self.run_command('env unset appa --all')
+        super(TestEnvCLI, self).tearDown()
+
+    def assertEnvInContainer(self, app_or_service, var, value):
+        env = self.exec_in_container(app_or_service, 'printenv {}'.format(var)).rstrip()
+        self.assertEqual(env, value)
+
+    def assertEnvNotInContainer(self, app_or_service, var):
+        env = self.exec_in_container(app_or_service, 'printenv {}'.format(var)).rstrip()
+        self.assertEqual(env, '')
```
Comment (thieman) [inline tests/integration/cli/env_test.py]: These should live in the testcase with the other asserts

File: tests/integration/cli/env_test.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        self.run_command('env set appa pitcher pedro')
+        self.run_command('up --no-pull')
+        self.assertEnvInContainer('appa', 'pitcher', 'pedro')
+
+    def test_env_overrides_spec_environment(self):
+        self.run_command('up --no-pull')
+        self.assertEnvInContainer('appa', 'SPEC_VALUE', 'spec-specified-value')
+        self.run_command('env set appa SPEC_VALUE new-value')
+        self.run_command('up --no-pull')
+        self.assertEnvInContainer('appa', 'SPEC_VALUE', 'new-value')
```
Comment (thieman) [inline tests/integration/cli/env_test.py]: Can you rewrite this to only need to do one up, instead specifying two different env vars and only overwriting one?

File: tests/integration/cli/env_test.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        self.assertEnvInContainer('appa', 'SPEC_VALUE', 'spec-specified-value')
+
+    def test_unset_one(self):
+        self.run_command('env set appa pitcher pedro')
+        self.run_command('env set appa SPEC_VALUE new-value')
+        self.run_command('up --no-pull')
+        self.run_command('env unset appa SPEC_VALUE')
+        self.run_command('up --no-pull')
+        self.assertEnvInContainer('appa', 'pitcher', 'pedro')
+        self.assertEnvInContainer('appa', 'SPEC_VALUE', 'spec-specified-value')
```
Comment (thieman) [inline tests/integration/cli/env_test.py]: Can these two also be made to only do `up` once?

File: dusty/cli/env.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+Usage:
+  env list [<app_or_service>]
+  env set <app_or_service> (<var_name> <value> | --file <local_file>)
+  env unset <app_or_service> (--all | <var_name>)
+
+Commands:
+  list        List all environment variables and their current values.
+  set         Set a variable name to a new value for the given app or service.
+  unset       Unset a variable for the given app or service.
+"""
```
Comment (thieman) [inline dusty/cli/env.py]: Throw this in the usage docs. Also add this feature to changelog.

File: tests/integration/cli/env_test.py (hunk 5 above)
Comment (jsingle) [inline tests/integration/cli/env_test.py]: hmm we often do.  I'm also unsure why

File: dusty/cli/env.py (hunk 9 above)
Comment (jsingle) [inline dusty/cli/env.py]: Done

**Comment by jsingle** [comment]
I'm thinking that integration test failure is not related to this.  Mac slave might be sad b/c zach pushed a non-rebased PR

File: docs/changelog.md [hunk 10]
```diff
@@ -7,6 +7,7 @@
     * The setup override flag `--boot2docker_vm_memory` has been renamed to `--vm_memory`
 * **New**
     * Managed repos are now updated in parallel during `dusty up` and `dusty test`.
+    * New command `dusty env` can set environment variable overrides.
```
Comment (thieman) [inline docs/changelog.md]: Flesh this out a little

**Comment by thieman** [comment]
LGTM, please squash

**Pull request merged**